sw/host/opentitanlib/src/transport/hyperdebug/mod.rs

```diff
@@ -0,0 +1,320 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::Result;
+
+use std::cell::RefCell;
+use std::collections::hash_map::Entry;
+use std::collections::HashMap;
+use std::fs;
+use std::path::{Path, PathBuf};
+use std::rc::Rc;
+
+use thiserror::Error;
+
+use crate::io::gpio::GpioPin;
+use crate::io::spi::Target;
+use crate::io::uart::Uart;
+use crate::transport::{Capabilities, Capability, Transport, TransportError};
+use crate::util::usb::UsbBackend;
+
+pub mod gpio;
+pub mod uart;
+pub mod spi;
+
+/// Implementation of the Transport trait for HyperDebug based on the
+/// Nucleo-L552ZE-Q.
+pub struct Hyperdebug {
+    spi_names: HashMap<String, u8>,
+    spi_interface: BulkInterface,
+    uart_ttys: HashMap<String, PathBuf>,
+    inner: Rc<Inner>,
+}
+
+/// Index of a single USB "interface", with its associated IN and OUT
+/// endpoints.  Used to instantiate e.g. SPI trait.
+#[derive(Copy, Clone)]
+pub struct BulkInterface {
+    interface: u8,
+    in_endpoint: u8,
+    out_endpoint: u8,
+}
+
+/// Internal state of the Hyperdebug struct, this struct is reference counted such that Gpio,
+/// Spi and Uart sub-structs can all refer to this shared data, which is guaranteed to live on,
+/// even if the caller lets the outer Hyperdebug struct run out of scope.
+pub struct Inner {
+    console_tty: PathBuf,
+    usb_device: RefCell<UsbBackend>,
+    gpio: RefCell<HashMap<String, Rc<dyn GpioPin>>>,
+    spis: RefCell<HashMap<u8, Rc<dyn Target>>>,
+    uarts: RefCell<HashMap<PathBuf, Rc<dyn Uart>>>,
+}
+
+impl Hyperdebug {
+    pub const VID_GOOGLE: u16 = 0x18d1;
+    pub const PID_HYPERDEBUG: u16 = 0x520e;
+
+    /// Establish connection with a particular HyperDebug.
+    pub fn open(usb_vid: Option<u16>, usb_pid: Option<u16>, usb_serial: &Option<String>)
+                -> Result<Self> {
+        let device = UsbBackend::new(
+            usb_vid.unwrap_or(Self::VID_GOOGLE),
+            usb_pid.unwrap_or(Self::PID_HYPERDEBUG),
+            usb_serial)?;
+
+        let path = PathBuf::from("/sys/bus/usb/devices");
+
+        let mut console_tty: Option<PathBuf> = None;
+        let mut spi_interface: Option<BulkInterface> = None;
+        let mut uart_ttys: HashMap<String, PathBuf> = HashMap::new();
+
+        let config_desc = device.active_config_descriptor()?;
+        // Iterate through each USB interface, discovering e.g. supported UARTs.
+        for interface in config_desc.interfaces() {
+            for interface_desc in interface.descriptors() {
+                if let Some(idx) = interface_desc.description_string_index() {
```

Comment:
You can save on indentation by writing something like
```rust
let idx = match ... {
  Some(idx) => idx,
  None => continue,
};
```

Reply:
Done